clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -343,9 +350,30 @@ private KafkaAdminClient(AdminClientConfig config, String clientId, Time time, M
     }
 
     @Override
-    public void close() {
-        closed = true;
-        client.wakeup(); // Wake the thread, if it is blocked inside poll().
+    public void close(long waitTimeDuration, TimeUnit waitTimeUnit) {
+        long waitTimeMs = waitTimeUnit.toMillis(waitTimeDuration);
+        long now = time.milliseconds();
+        long newHardShutdownTimeMs = now + waitTimeMs;
+        long prev = INVALID_SHUTDOWN_TIME;
+        while (true) {
+            if (hardShutdownTimeMs.compareAndSet(prev, newHardShutdownTimeMs)) {
+                if (prev == INVALID_SHUTDOWN_TIME) {
+                    log.debug("{}: initiating close operation.", clientId);
+                } else {
+                    log.debug("{}: moving hard shutdown time forward.", clientId);
+                }
+                client.wakeup(); // Wake the thread, if it is blocked inside poll().
+                break;
+            }
+            prev = hardShutdownTimeMs.get();
+            if (prev < newHardShutdownTimeMs) {
+                log.debug("{}: hard shutdown time is already earlier than requested.", clientId);
+                newHardShutdownTimeMs = prev;
+                break;
+            }
+        }
+        long deltaMs = Math.max(0, newHardShutdownTimeMs - time.milliseconds());
+        log.debug("{}: waiting for the I/O thread to exit.  Hard shutdown in {} ms.", clientId, deltaMs);
```

Comment:
Seems like we only compute `deltaMs` for the `debug` call. Should we be using a `isDebugEnabled` guard?